DaemonSets should allow MaxSurge to minimize workload downtime during upgrade

smarterclayton opened the pull request:
Many infrastructure components require DaemonSets to place pods on
each node, but the current update strategies limit end users from
minimizing disruption during updates. It should be possible to
surge daemonsets if a user requests it to allow handoff from one
running pod to another, like Deployments.

Extracts the specific Surge portion from #1140.

liggitt commented on keps/sig-apps/20200302-surge-on-daemonsets.md:
```diff
@@ -0,0 +1,148 @@
+
+The design of Deployment rolling updates introduced the surge concept, and the initial design for DaemonSet updates considered the implications of adding the Surge strategy later (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/apps/daemonset-update.md#future-plans).
+
+We would add `MaxSurge *intstr.IntOrString` to the RollingUpdate deployment. It would have a default value of 0, preserving current behavior. We would allow MaxUnavailable to be 0 when MaxSurge is set.
```
what are the current validation requirements around MaxUnavailable? we would need to preserve those for at least one release for newly created DaemonSets and for updated DaemonSets where the existing object fit within those validation requirements

liggitt replied:
current validation for MaxUnavailable:
```
// This cannot be 0.
```

smarterclayton replied:
Yes, I would assume that would be beta to GA transition, unless you think we'd do a longer than one release wait.

liggitt commented on keps/sig-apps/20200302-surge-on-daemonsets.md:
```diff
@@ -0,0 +1,148 @@
+
+The last category category is the most difficult to achieve and generally combines categories 1 and 2 along with careful tuning. Networking plugins that provide pod network capability may have one or more daemon processes that are desirable to deliver containerized, but any disruption to those critical pods may impact other workloads. In most cases, the capability to overlap execution provided by ...
+
+### Risks and Mitigations
```
if this is proposing allowing additional values for MaxUnavailable, describe the phased rollout that ensures we never allow DaemonSet data to persist which an active API server in the cluster would consider invalid

smarterclayton replied:
Covered, let me know what you think.

liggitt commented on keps/sig-apps/20200302-surge-on-daemonsets.md:
```diff
@@ -0,0 +1,148 @@
+
+In the event a rollout cannot proceed due to hitting the MaxSurge limit (due to any condition, whether scheduling, new pods not going ready) the controller should pause creating new pods until conditions change.
+
+DaemonSet pods are slightly more constrained than Deployments when it comes to scheduling issues since each pod is tied to a single node, so it is worth describing exactly how surge pods that violate same node constraints would be handled consistent with Deployments. The most common conflict is use of HostPort within the pod spec across two versions, which would prevent the second pod from landin ...
```
> It is user error to specify impossible scheduling constraints

testing should exercise recovering from scenarios like this... if I specify a surge and a hostPort, and get 30% of my nodes stuck with a DS pod pending scheduling, switching rollout strategies or removing the hostPort should start a new rollout and unstick my DS, right?

smarterclayton replied:
Yes.

liggitt commented on keps/sig-apps/20200302-surge-on-daemonsets.md:
```diff
@@ -0,0 +1,148 @@
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
```
does PDB need to be updated for this as well?

smarterclayton replied:
I can add some comments in here, but drain's goal is to remove pods from the node.  If you choose to drain daemonsets, you have to match whatever the PDB constraint was, and since daemonsets ignore unschedulable you don't really get the right outcome from a PDB.  You drain only if the requisite PDB reqt is met, but then immediately after drain succeeds the pod comes right back and PDB goes back up, so a parallel drain is going to potentially see you fail.  If you have a large drain then yes PDB would limit you (as long as the node objects were still there).

smarterclayton replied:
added implications section

liggitt commented:
Overall, this looks consistent with deployments. DaemonSets' scheduling constraints make it more likely a scheduling conflict would be hit if the pod used any exclusive resource (like a hostPort), but since it is also possible to specify a deployment that 1) uses a hostPort and 2) specifies a nodeSelector that constrains pods to a set of nodes saturated by existing pods from the deployment, I would expect us to be able to use similar logic to recover rollouts from impossible user-specified constraints.

tnozicka commented on keps/sig-apps/20200302-surge-on-daemonsets.md:
```diff
@@ -0,0 +1,148 @@
+	// can have their a new pod created before the old pod is marked as deleted.
+  // The update starts by launching new pods on 30% of nodes. Once an updated
+  // pod is available (Ready for at least minReadySeconds) the old DaemonSet pod
+  // on that node is marked deleted. If the old pod becomes unavailable for any
```
I think I'd like to see a test case for old and new pods going unready during the rollout.

tnozicka commented on keps/sig-apps/20200302-surge-on-daemonsets.md:
```diff
@@ -0,0 +1,148 @@
+  // pod is available (Ready for at least minReadySeconds) the old DaemonSet pod
+  // on that node is marked deleted. If the old pod becomes unavailable for any
+  // reason (Ready transitions to false, is evicted, or is drained) an updated
+  // pod is immediatedly created on that node without considering surge limits.
```
I see that this is following the availability reconciliation. An edge case is the resource consumption that could sky rocket on readiness blip. Like running 3 masters on one VM and this immediately doubles resources for intensive DS, although I've specified `MaxSurge: 1`.

smarterclayton replied:
Only if you have an upgrade pending.  If you select surge != 0 you are allowing double resource consumption per node.  So you were going to double up resources on that node at some point in time `t` after `t0` (when you updated), all that the blip does is move it forward a bit faster.  If you have no upgrade pending, then we don't launch another pod.  

Agree the docs should call out resource consumption mindset.

smarterclayton commented on keps/sig-apps/20200302-surge-on-daemonsets.md:
```diff
@@ -0,0 +1,148 @@
+	// The maximum number of DaemonSet pods that can be unavailable during the
+	// update. Value can be an absolute number (ex: 5) or a percentage of total
+	// number of DaemonSet pods at the start of the update (ex: 10%). Absolute
+	// number is calculated from percentage by rounding up.
```
Good point, we should be consistent with deployments unless there is a significant motivating reason (bar should be high).

smarterclayton replied:
commented inline with fencepost rule.

smarterclayton commented:
Updated comments.

smarterclayton commented:
@tnozicka we should probably bring this up on next sig-apps and get extra eyes on it

janetkuo commented:
I put this KEP in the discussion topic for the next SIG Apps meeting (June 29th.)
https://github.com/kubernetes/community/tree/master/sig-apps#meetings

smarterclayton commented:
Comments from sig-apps

* talk about how this would be impacted / could impact service topology work
* clarify connection to statefulset surge proposal
* discuss implications of validation w.r.t. HostPort being rejected with (at start) and implications for upgrade (if this was set, validation should not fire to not block updates / storage migrations)

Will check the kep status is correct and add those updates

krmayankk commented on keps/sig-apps/20200302-surge-on-daemonsets.md:
```diff
@@ -0,0 +1,163 @@
+
+The design of Deployment rolling updates introduced the surge concept, and the initial design for DaemonSet updates considered the implications of adding the Surge strategy later (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/apps/daemonset-update.md#future-plans).
+
+We would add `MaxSurge *intstr.IntOrString` to the RollingUpdate deployment. It would have a default value of 0, preserving current behavior. We would allow MaxUnavailable to be 0 when MaxSurge is set.
```
nit: `RollingUpdate daemonset `

krmayankk commented on keps/sig-apps/20200302-surge-on-daemonsets.md:
```diff
@@ -0,0 +1,163 @@
+
+* HostPort use when MaxSurge != 0
+
+However in general we do not perform this validation in other contexts, and its novelty probably does not justify the extra complexity.
```
how about conditions or events which show up in describe that warn the user of possible issues when hostPath is set and maxsurge != 0

smarterclayton replied:
After discussion in the sig meeting i updated to add validation rules that prevent this scenario.

krmayankk commented on keps/sig-apps/20200302-surge-on-daemonsets.md:
```diff
@@ -0,0 +1,163 @@
+Infrastructure invoked by Kubernetes components (CRI, CNI, CSI) can usually fall within the first category and may require some coordination from the invoking process to minimize downtime. For instance, the Kubelet may retry certain types of CSI errors transparently to mitigate brief disruption to a CSI plugin. Or the container runtime may retry certain CNI errors if the plugin is not available.
+
+The second category of workload requires some coordination between the old and new container - for instance, reusing a host volume and checking for file locking on shared resources, or using the SO_REUSEPORT option to start listening on an interface and share old and new traffic. In general the workload author is assumed to understand how to minimize disruption and Kubernetes is only giving them  ...
+
```
we should allow the user controlling the overlap window  between old and new pods. Would that be through readiness probe ?

smarterclayton replied:
Yes, ready controls gating.  Updated.

kikisdeliveryservice reviewed (commented):
This is using an old KEP format. Please see : https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template and update accordingly.

smarterclayton commented:
I have updated this with all feedback from sig apps, updated to the new kep format, and put it into implementable (given last feedback).  PTAL @kow3ns @tnozicka @janetkuo - sorry for the long delay.

soltysh commented on keps/sig-apps/1591-daemonset-surge/README.md:
```diff
@@ -0,0 +1,138 @@
+  // pod is available (Ready for at least minReadySeconds) the old DaemonSet pod
+  // on that node is marked deleted. If the old pod becomes unavailable for any
+  // reason (Ready transitions to false, is evicted, or is drained) an updated
+  // pod is immediatedly created on that node without considering surge limits.
```
Nit: formatting

soltysh commented on keps/sig-apps/1591-daemonset-surge/README.md:
```diff
@@ -0,0 +1,138 @@
+
+DaemonSet pods are slightly more constrained than Deployments when it comes to scheduling issues since each pod is tied to a single node, so it is worth describing exactly how surge pods that violate same node constraints would be handled consistent with Deployments. The most common conflict is use of HostPort within the pod spec across two versions, which would prevent the second pod from landin ...
+
+In order to reduce confusion for new users, we will start by rejecting HostPort use in daemonset when surge is MaxSurge is non-zero. A user will not be able to update a daemonset to MaxSurge != 0 if HostPort is set, or update a HostPort if MaxSurge is set, without receiving a validation error. If the MaxSurge feature gate is off, the validation rule is bypassed, and a user who turns off the gate, ...
```
we will start by rejecting HostPort use in daemonset when ~~surge is~~ MaxSurge is non-zero.

soltysh commented on keps/sig-apps/1591-daemonset-surge/README.md:
```diff
@@ -0,0 +1,138 @@
+
+The second category of workload requires some coordination between the old and new container - for instance, reusing a host volume and checking for file locking on shared resources, or using the SO_REUSEPORT option to start listening on an interface and share old and new traffic. In general the workload author is assumed to understand how to minimize disruption and Kubernetes is only giving them  ...
+
+The last category category is the most difficult to achieve and generally combines categories 1 and 2 along with careful tuning. Networking plugins that provide pod network capability may have one or more daemon processes that are desirable to deliver containerized, but any disruption to those critical pods may impact other workloads. In most cases, the capability to overlap execution provided by ...
```
The last category ~~category~~ is

soltysh commented on keps/sig-apps/1591-daemonset-surge/README.md:
```diff
@@ -0,0 +1,138 @@
+
+### Risks and Mitigations
+
+The primary risk is a bug in the implementation of the controller that causes excessive pod creations or deletions, as we have experienced during previous enhancements to workload controllers. The best mitigation for that is scenario unit testing to ensure the update strategy is stable and general purpose stress e2e testing of the controller.
```
The best mitigation for that scenario is unit (is and scenario swapped)

soltysh reviewed (commented):
A few nits, but most importantly you're missing PRR section

tnozicka reviewed (approved):
/approve

smarterclayton commented:
Add PRR section and updated with comments.

soltysh commented on keps/sig-apps/1591-daemonset-surge/README.md:
```diff
@@ -0,0 +1,343 @@
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name:
+    - Components depending on the feature gate:
```
Nit: You need to copy here values from `kep.yaml`

smarterclayton replied:
Ah.  Yuck.

soltysh replied:
Yeah, I know :sweat:

soltysh reviewed (approved):
/lgtm

/assign @deads2k 
for PRR review

deads2k commented on keps/sig-apps/1591-daemonset-surge/README.md:
```diff
@@ -0,0 +1,343 @@
+* **What are the SLIs (Service Level Indicators) an operator can use to determine
+the health of the service?**
+  - [ ] Metrics
+    - Metric name:
```
I'd really like to see a metric about how frequently this feature is used.  It helps cluster-admins (and potentially distributors), know how broadly used the feature is.

krmayankk replied:
How are these kinds of metrics currently tracked ? I once saw dashboard for api usage, but dont know how it was created in one of the sigs meetings

deads2k commented on keps/sig-apps/1591-daemonset-surge/README.md:
```diff
@@ -0,0 +1,343 @@
+* **What are the SLIs (Service Level Indicators) an operator can use to determine
+the health of the service?**
+  - [ ] Metrics
+    - Metric name:
```
Is there a "surging stuck" metric?  I'm imaging cases where a host network port is reused or some other condition where the surging pod doesn't run right.  It seems like a more likely risk for daemonsets than for other workload types.

deads2k commented on keps/sig-apps/1591-daemonset-surge/README.md:
```diff
@@ -0,0 +1,343 @@
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **Does this feature depend on any specific services running in the cluster?**
```
as I recall, daemonsets are a fairly tight integration between controller-manager and scheduler.  Is this fully contained to a single binary?

deads2k commented:
Since this is alpha, I consider the PRR questions beyond the "required for alpha" section as non-blocking.  I do think some additional metrics could help in evaluating the alpha level as well.

mattfarina commented:
/approve

Cleared with @kow3ns first

smarterclayton commented:
API PR is up in https://github.com/kubernetes/kubernetes/pull/96375
